internal/notification/handlers/notification_worker.go

```diff
@@ -0,0 +1,494 @@
+			NotificationType:              e.NotificationType,
+			MessageType:                   e.MessageType,
+			UnverifiedNotificationChannel: e.UnverifiedNotificationChannel,
+			Args:                          e.Args,
+			AggregateID:                   e.AggregateID,
+			AggregateResourceOwner:        e.AggregateResourceOwner,
+			IsOTP:                         e.IsOTP,
+		},
+		BackOff:    backoff,
+		NotifyUser: notifyUser,
+	}
+}
+
+func (w *NotificationWorker) schedule(ctx context.Context, workerID int, retry bool) {
+	t := time.NewTimer(0)
+
+	for {
+		select {
+		case <-ctx.Done():
+			t.Stop()
+			w.log(workerID, retry).Info("scheduler stopped")
+			return
+		case <-t.C:
+			instances, err := w.queryInstances(ctx, retry)
+			w.log(workerID, retry).OnError(err).Error("unable to query instances")
+
+			w.triggerInstances(call.WithTimestamp(ctx), instances, workerID, retry)
+			if retry {
+				t.Reset(w.config.RetryRequeueEvery)
+				continue
+			}
+			t.Reset(w.config.RequeueEvery)
+		}
+	}
+}
+
+func (w *NotificationWorker) log(workerID int, retry bool) *logging.Entry {
+	return logging.WithFields("notification worker", workerID, "retries", retry)
+}
+
+func (w *NotificationWorker) queryInstances(ctx context.Context, retry bool) ([]string, error) {
+	if w.config.HandleActiveInstances == 0 {
+		return w.existingInstances(ctx)
+	}
+
+	query := eventstore.NewSearchQueryBuilder(eventstore.ColumnsInstanceIDs).
+		AwaitOpenTransactions().
+		AllowTimeTravel().
+		CreationDateAfter(w.now().Add(-1 * w.config.HandleActiveInstances))
+
+	maxAge := w.config.RequeueEvery
+	if retry {
+		maxAge = w.config.RetryRequeueEvery
+	}
+	return w.es.InstanceIDs(ctx, maxAge, false, query)
+}
+
+func (w *NotificationWorker) existingInstances(ctx context.Context) ([]string, error) {
+	ai := existingInstances{}
+	if err := w.es.FilterToQueryReducer(ctx, &ai); err != nil {
+		return nil, err
+	}
+
+	return ai, nil
+}
+
+func (w *NotificationWorker) triggerInstances(ctx context.Context, instances []string, workerID int, retry bool) {
+	for _, instance := range instances {
+		instanceCtx := authz.WithInstanceID(ctx, instance)
+
+		err := w.trigger(instanceCtx, workerID, retry)
+		w.log(workerID, retry).WithField("instance", instance).OnError(err).Info("trigger failed")
+	}
+}
+
+func (w *NotificationWorker) trigger(ctx context.Context, workerID int, retry bool) (err error) {
+	txCtx := ctx
+	if w.config.TransactionDuration > 0 {
+		var cancel, cancelTx func()
+		txCtx, cancelTx = context.WithCancel(ctx)
+		defer cancelTx()
+		ctx, cancel = context.WithTimeout(ctx, w.config.TransactionDuration)
+		defer cancel()
+	}
+	tx, err := w.client.BeginTx(txCtx, nil)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = database.CloseTransaction(tx, err)
+	}()
+
+	eventType := eventstore.EventType(notification.RequestedType)
+	exclude := []eventstore.EventType{notification.RetryRequestedType, notification.CanceledType, notification.SentType}
+	if retry {
```

Comment:
I would personally prefer on splitting this into two functions instead of a condition.